tests/sources/test_cppcheck.py

```diff
@@ -101,57 +167,35 @@ def setUp(self):
-        """Test Webots with Cppcheck."""
-        command = self.cppcheck + ' --enable=warning,style,performance,portability --inconclusive --force -q'
-        command += ' -j %s' % str(multiprocessing.cpu_count())
-        command += ' --inline-suppr --suppress=invalidPointerCast --suppress=useStlAlgorithm --suppress=uninitMemberVar '
-        command += ' --suppress=noCopyConstructor  --suppress=noOperatorEq'
-        # command += ' --xml '  # Uncomment this line to get more information on the errors
-        command += ' --output-file=' + self.reportFilename
-        for include in self.includeDirs:
-            command += ' -I\"' + os.path.normpath(self.WEBOTS_HOME + '/' + include) + '\"'
-        for source in self.skippedDirs:
-            command += ' -i\"' + os.path.normpath(self.WEBOTS_HOME + '/' + source) + '\"'
-        for source in self.sourceDirs:
-            command += ' \"' + os.path.normpath(self.WEBOTS_HOME + '/' + source) + '\"'
-        self.run_cppcheck(command)
-
-    def test_projects_with_cppcheck(self):
-        """Test projects with Cppcheck."""
+        files_diff = os.path.join(self.WEBOTS_HOME, 'tests', 'sources', 'files_diff.txt')
         command = self.cppcheck + ' --enable=warning,style,performance,portability --inconclusive --force -q '
         command += '--inline-suppr --suppress=invalidPointerCast --suppress=useStlAlgorithm -UKROS_COMPILATION '
         # command += '--xml '  # Uncomment this line to get more information on the errors
-        command += '--std=c++03 --output-file=' + self.reportFilename
-        for source in self.projectsSkippedDirs:
-            command += ' -i\"' + os.path.normpath(self.WEBOTS_HOME + '/' + source) + '\"'
-        for source in self.projectsSourceDirs:
-            command += ' \"' + os.path.normpath(self.WEBOTS_HOME + '/' + source) + '\"'
+        command += '--std=c++03 --output-file=\"' + self.reportFilename + '\"'
+        if os.path.isfile(files_diff):
+            added = False
+            file = open(files_diff, 'r')
+            for line in file:
+                line = line.strip()
+                extension = os.path.splitext(line)[1][1:].lower()
+                if extension not in self.extensions:
+                    continue
+                for projectsSourceDir in projectsSourceDirs:
+                    if line.startswith(projectsSourceDir):
+                        for projectSkippedDir in projectsSkippedDirs:
+                            if not line.startswith(projectSkippedDir):
+                                command += ' \"' + line + '\"'
+                                added = True
+                                break
+            file.close()
+            if not added:
+                return
+        else:
+            for source in projectsSkippedDirs:
+                command += ' -i\"' + source + '\"'
+            for source in projectsSourceDirs:
+                command += ' \"' + source + '\"'
+        os.chdir(self.WEBOTS_HOME)
```

Comment:
Is this really required? This looks dangerous and I am not sure to understand why you need it?

Reply:
That's because I pass all the files and folder to `cppcheck` as relative to `WEBOTS_HOME` and not absolute. This is very useful to reduce the size of the command line which can be very long and sometimes exceeds the maximum length allowed on Windows, e.g., 8192 chars.

Reply:
To improve this, I can save the previous `cwd` and restore at after the command has been issued.

Reply:
In most of the cases the tests are already run in the correct folder because we call them with `python -m unittest discover -s ...`. But I can indeed reproduce the issue when calling directly this test.

I think it would indeed be safer to restore the `cwd`, I am a bit afraid that otherwise it can 'pollute' the next test when launched with `unittest`.